apple/internal/xcarchive.bzl

```diff
@@ -0,0 +1,92 @@
+"""
+Rule for packaging a .ipa or .app into a .xcarchive.
+"""
+
+load(
+    "@build_bazel_rules_apple//apple:providers.bzl",
+    "AppleBundleInfo",
+    "AppleDsymBundleInfo",
+)
+
+def _xcarchive_impl(ctx):
+    """
+    Implementation for xcarchive.
+
+    This rule uses the providers from the app target to re-package an .ipa or .app bundle into a
+    .xcarchive. The .xcarchive is a directory that contains the app bundle, dSYM and other metadata.
+
+    It calls out to a binary to do the actual re-packaging.
+    """
+    bundle_info = ctx.attr.app[AppleBundleInfo]
+    dsym_info = ctx.attr.app[AppleDsymBundleInfo]
+    xcarchive = ctx.actions.declare_directory("%s.xcarchive" % bundle_info.bundle_name)
+
+    arguments = ctx.actions.args()
+    arguments.add("--info_plist", bundle_info.infoplist.path)
+    arguments.add("--archive", bundle_info.archive.path)
+    arguments.add("--output", xcarchive.path)
+
+    for dsym in dsym_info.direct_dsyms:
+        arguments.add("--dsym", dsym.path)
+
+    ctx.actions.run(
+        inputs = depset([
+            bundle_info.archive,
+            bundle_info.infoplist,
+        ] + dsym_info.direct_dsyms),
+        outputs = [xcarchive],
+        executable = ctx.executable._make_xcarchive,
+        arguments = [arguments],
+        mnemonic = "XCArchive",
+    )
+
+    return [
+        DefaultInfo(files = depset([xcarchive])),
+    ]
+
+xcarchive = rule(
+    implementation = _xcarchive_impl,
+    attrs = {
+        "app": attr.label(
```

Comment:
Maybe name this `bundle` since it could be a non-app.

Reply:
Renamed to `bundle`